### PR TITLE
new vm 1.5.12

### PR DESCRIPTION
**Pull request by sasurobert**
## Reasoning behind the pull request
- 
- 
- 
  
## Proposed changes
- 
- 
- 

## Testing procedure
- standard system test
- on previous versions (everything starting from `v1.5.13-dev-config` up until this branch) SC deployments & upgraded were allowed to any .wasm regardless if the `init` and `upgrade` functions were defined or not. Starting with this branch the deploy and upgrade will need to have .wasm files that will contain `init` / `upgrade` functions. Otherwise the deploy/upgrade transactions will fail.
 
## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -21,7 +21,7 @@ require (
 	github.com/multiversx/mx-chain-scenario-go v1.2.1
 	github.com/multiversx/mx-chain-storage-go v1.0.13
 	github.com/multiversx/mx-chain-vm-common-go v1.5.5
-	github.com/multiversx/mx-chain-vm-go v1.5.10
+	github.com/multiversx/mx-chain-vm-go v1.5.13-0.20231005091322-28dffccc749e
```
Comment (iulianpascalau) [inline go.mod:24]: proper release after testing

File: integrationTests/vm/txsFee/dynamicGasCost_test.go [hunk 2]
```diff
@@ -47,15 +47,17 @@ func TestDynamicGasCostForDataTrieStorageLoad(t *testing.T) {
 	dataTrieInstance := getAccountDataTrie(t, testContext, scAddress)
 	trieKeysDepth := getTrieDepthForKeys(t, dataTrieInstance, keys)
 
+	// Question for review: is this all right?
```
Comment (iulianpascalau) [inline integrationTests/vm/txsFee/dynamicGasCost_test.go:50]: yes, we can remove the comment

File: integrationTests/vm/txsFee/multiShard/relayedScDeploy_test.go [hunk 3]
```diff
@@ -71,13 +71,10 @@ func TestRelayedSCDeployShouldWork(t *testing.T) {
...
 	accumulatedFees = testContextInner.TxFeeHandler.GetAccumulatedFees()
-	require.Equal(t, big.NewInt(8490), accumulatedFees)
+	require.Equal(t, big.NewInt(19600), accumulatedFees)
 
 	txs := testContextInner.GetIntermediateTransactions(t)
 
 	scr := txs[0]
-	utils.ProcessSCRResult(t, testContextRelayer, scr, vmcommon.Ok, nil)
-
-	expectedBalanceRelayer = big.NewInt(28440)
```
Comment (iulianpascalau) [inline integrationTests/vm/txsFee/multiShard/relayedScDeploy_test.go:81]: can we put back this check even if we need to adjust the expected balance?

File: integrationTests/vm/wasm/delegation/delegation_test.go [hunk 4]
```diff
@@ -33,39 +32,6 @@ var NewBalance = wasm.NewBalance
 var NewBalanceBig = wasm.NewBalanceBig
 var RequireAlmostEquals = wasm.RequireAlmostEquals
 
-func TestDelegation_Upgrade(t *testing.T) {
```
Comment (iulianpascalau) [inline integrationTests/vm/wasm/delegation/delegation_test.go:36]: why did we remove this test?

File: integrationTests/vm/wasm/upgrades/upgrades_test.go [hunk 5]
```diff
@@ -126,40 +126,6 @@ func TestUpgrades_ParentAndChildContracts(t *testing.T) {
 	require.Equal(t, uint64(42), context.QuerySCInt("getUltimateAnswer", [][]byte{}))
 }
 
-func TestUpgrades_UpgradeDelegationContract(t *testing.T) {
```
Comment (iulianpascalau) [inline integrationTests/vm/wasm/upgrades/upgrades_test.go:129]: why did we remove this test?

**Review by gabi-vuls — DISMISSED** [review]
Normal allin test: v1.5.13-dev-config-8c685c8e00 -> vm1.5.12-8e00be98a6

--- Specific errors ---

block hash does not match 9604
wrong nonce in block 3437
miniblocks does not match 0
num miniblocks does not match 0
miniblock hash does not match 0
block bodies does not match 0
receipts hash missmatch 0

/------/


--- Statistics ---

Nr. of all ERRORS:  0
Nr. of all WARNS:  247
Nr. of new ERRORS:  0
Nr. of new WARNS:  11
Nr. of PANICS:  0

/------/

--- ERRORS ---

/------/

--- WARNINGS ---

File: integrationTests/vm/wasm/upgrades/upgrades_test.go (hunk 5 above)
Comment (andreibancioiu) [reply integrationTests/vm/wasm/upgrades/upgrades_test.go:129]: It was, somehow, a duplicate (we already have a few tests that do test the upgrade flow). Plus, the delegation contract is the one written in Rust, plus it's a bit more difficult to find a version to be used in tests - then to add a version to the binary (filename) etc. So, removed for less testing bureaucracy, without actually harming the confidence in the upgrade flow.

File: integrationTests/vm/wasm/delegation/delegation_test.go (hunk 4 above)
Comment (andreibancioiu) [reply integrationTests/vm/wasm/delegation/delegation_test.go:36]: Explained in comment below. It's a bit hard to obtain a variant of the delegation contract that satisfies the `upgrade` requirement - without additional GitHub / cross-repo / versioning bureaucracy. Confidence of tests remains the same.

**Review by gabi-vuls — APPROVED** [review]
ormal allin test: v1.5.13-dev-config-8c685c8e00 -> vm1.5.12-81dad39942

--- Specific errors ---

block hash does not match 1742
wrong nonce in block 806
miniblocks does not match 0
num miniblocks does not match 0
miniblock hash does not match 0
block bodies does not match 0
receipts hash missmatch 0

/------/


--- Statistics ---

Nr. of all ERRORS:  0
Nr. of all WARNS:  159
Nr. of new ERRORS:  0
Nr. of new WARNS:  1
Nr. of PANICS:  0

/------/

--- ERRORS ---

/------/

--- WARNINGS ---

File: vm/systemSmartContracts/governance.go [hunk 6]
```diff
@@ -609,7 +612,7 @@ func (g *governanceContract) closeProposal(args *vmcommon.ContractCallInput) vmc
 	}
 
 	currentEpoch := g.eei.BlockChainHook().CurrentEpoch()
-	if uint64(currentEpoch) < generalProposal.EndVoteEpoch {
+	if uint64(currentEpoch) <= generalProposal.EndVoteEpoch {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/governance.go:615]: nothing to do with this PR. We should avoid these types of hot fixes

File: integrationTests/vm/txsFee/multiShard/relayedScDeploy_test.go (hunk 3 above)
Comment (sasurobert) [reply integrationTests/vm/txsFee/multiShard/relayedScDeploy_test.go:81]: it was a non sense.

**Review by iulianpascalau — DISMISSED** [review]

**Review by andreibancioiu — APPROVED** [review]

**Review by gabi-vuls — DISMISSED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]

**Review by andreibancioiu — APPROVED** [review]

**Pull request merged**